### ISSUE TITLE
Content security policy - unsafe-eval

### ISSUE THREAD

**Issue by lol768** [MiniProfiler/dotnet#235]
Hey,

I'm using MiniProfiler in an application with a pretty restrictive `Content-Security-Policy`. At present, the `script-src` directive does not permit the `unsafe-eval` source which breaks the MiniProfiler UI.

Would it be possible to change the templates to avoid this sort of code?

```
Error: call to Function() blocked by CSP    includes.js:965:4245
Content Security Policy: The page’s settings blocked the loading of a resource at self (“default-src https://app.local:443”). Source: call to eval() or related function blocked by CSP.     includes.js:965
```

**Comment by NickCraver** [comment]
@lol768 That's jQuery 1.x, I'd been considering it for a while but this issue enough motivation to do so. 

I'll take a look at upgrading soon as I can this week, bumping us to 3.x. It's not a trivial change as it'll probably break some things on top (like templates) that I need to fix, but I agree this needs to happen. I'll update here when I get it working on jQuery 3.x.

**Comment by NickCraver** [comment]
Alright if you can give MyGet feed packages a try, the latest version upgraded to jQuery 3 and no longer has `eval`. Any chance you can test them out and give me feedback there? Testing here looks good; surprisingly no impact from the core upgrade.

**Comment by lol768** [comment]
Hey Nick,

I'm on version `4.0.0-alpha9-00207` and can confirm I can see `jQuery v3.2.1` in the `includes.js` file.

Unfortunately the issue persists with this build. Whilst `eval()` isn't called directly, `Function()` is called (and considered a related function to block) and the browser refuses to allow this:

```
Error: call to Function() blocked by CSP    includes.js:965:4245
```

I strongly suspect `jquery-tmpl` is to blame here. Is there any way these templates can be precompiled or migrated to use a library that doesn't use these unsafe constructs?

Cheers.

**Comment by NickCraver** [comment]
It's possible, but it's a significant change that'd require a lot of time to get back on par and forever increase maintenance cost. I'm not per se against it, but doubt it's worth it.

I'd be happy to hear recommendations on lower-maintenance cost pre-compilation options here, as I'm unfamiliar with what's out there at the moment.

**Comment by lol768** [comment]
I've used https://developers.google.com/closure/templates before. You're right in that it's an additional build step, but it would bring with it a few advantages:

* Better performance, since templates are compiled at build time and not at runtime.
* Closure Templates is actually maintained, whereas it looks like `jquery-tmpl` is no longer being developed. It's widely deployed, and is built to be secure (wrt cross-site-scripting at least).
* No need to allow `unsafe-eval` when showing the profiler, which _is_ a security risk. I've seen XSS vulns on sites that allow the user to influence information passed to jQuery's `append` function which would have been prevented with a proper CSP that didn't whitelist the `unsafe-eval` source.

You can make the argument that the profiler will only be visible e.g. on a development instance or for certain types of users and therefore the risk is decreased because the restrictive CSP can still be deployed for the majority of users in production. As a temporary workaround, I've got some logic that alters the `Content-Security-Policy` if the profiler is visible but this is additional complexity and a risk I'm not really happy with for the purpose of one library.

**Comment by dustinsoftware** [comment]
@lol768 FWIW we had to do the same thing on our website - the CSP gets altered if profiling mode is enabled for the current user, which only employee accounts can change. Since the CSP is a last-chance method to catch script injection attacks, and it will only be altered when someone is debugging the site, it isn't a big deal for us.

**Comment by lol768** [comment]
> @lol768 FWIW we had to do the same thing on our website - the CSP gets altered if profiling mode is enabled for the current user, which only employee accounts can change. Since the CSP is a last-chance method to catch script injection attacks, and it will only be altered when someone is debugging the site, it isn't a big deal for us.

If you're happy with the risk, then fair enough. At least for my application, the accounts with access to enable the profiler have permission to do a lot more damage. Users logged into these accounts are the ones for which an XSS would be most problematic and for which CSP (as a defense in depth measure of course) is important and as a result I'm not very happy with the status quo - especially when there are other reasons to switch to a better maintained templating engine.

MiniProfiler isn't alone by any stretch of the imagination here. There are a depressing number of JS libraries that will screw over your content security policy if you want them to work (Modernizr being another) which harms CSP adoption.

**Comment by Daniel15** [comment]
The UI for MiniProfiler is pretty simple... Could you just use raw DOM manipulation? Loading the whole of jQuery just to render MiniProfiler is pretty heavy.

**Comment by NickCraver** [comment]
@Daniel15 Could we? Yes, but it's not a "just" it's a lot of work and maintenance to get to where we are today - and jQuery isn't only used to render. To get rid of jQuery you'd need to do much more (and handle browser quirks of all those things). As I said I'm not against it, but it's costly and very low priority compared to everything else at the moment.

**Comment by lol768** [comment]
Thanks for sorting this, @NickCraver!

**Comment by NickCraver** [comment]
@lol768 I forgot to respond here because the auto-close, doh! Welcome! :)
